### PR TITLE
Fix flush read

### PR DESCRIPTION
**Pull request by muxi**
Fix some issues in flush read mechanism, and improve the way to reject authority header then cancel the call.

### CODE REVIEW

File: src/core/ext/transport/cronet/transport/cronet_transport.c [hunk 1]
```diff
@@ -155,7 +156,9 @@ struct op_state {
   bool flush_read;
   bool flush_cronet_when_ready;
   bool pending_write_for_trailer;
-  bool unprocessed_send_message;
+  bool pending_send_message;
+  bool pending_recv_trailing_metadata;
```
Comment (makdharma) [inline src/core/ext/transport/cronet/transport/cronet_transport.c]: add "// User has requested trailing metadata or status". Also explain the condition for flushing.

File: src/core/ext/transport/cronet/transport/cronet_transport.c [hunk 2]
```diff
@@ -248,11 +251,28 @@ static const char *op_id_string(enum e_op_id i) {
...
+static void null_and_maybe_free_read_buffer(stream_obj *s) {
   if (s->state.rs.read_buffer &&
       s->state.rs.read_buffer != s->state.rs.grpc_header_bytes) {
     gpr_free(s->state.rs.read_buffer);
-    s->state.rs.read_buffer = NULL;
+  }
+  s->state.rs.read_buffer = NULL;
+}
+
+static void maybe_flush_read(stream_obj *s) {
```
Comment (makdharma) [inline src/core/ext/transport/cronet/transport/cronet_transport.c]: // Add comment.

**Comment by muxi** [comment]
test this please

**Comment by muxi** [comment]
Test failures:
Basic Tests Linux - #10137
Basic Tests Mac - #10138
Tsan C++ - #10139

**Pull request merged**